---
title: "EXP2FLUX - Convert gene expression data to FBA flux boundaries"
author: "Daniel Osorio, Kelly Botero, Janneth Gonzalez and Andres Pinzon"
#date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to EXP2FLUX}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE, dev = "png",fig.width = 7.1, fig.height = 5, message = FALSE, warning = FALSE)
```
### Introduction to EXP2FLUX
Computational simulations of metabolism can help to predict the metabolic phenotype of an organism in response to different stimuli, through constraint-based modeling approaches. To recreate specific metabolic phenotypes and enhance the model predictive accuracy, several methods for the integration of transcriptomics data into constraint-based models have been proposed. The majority of available implemented methods are based on the discretization of data to incorporate constraints into the metabolic models via boolean logic representation, which reduces the accurate of physiological representations. The implemented methods for gene-expression data integration as continuous values are very few. The **exp2flux** package was designed as a tool to incorporate in a continuous way the gene-expression data as FBA flux limit in a metabolic model. Also, a function to calculate the differences between model fluxes in different metabolic scenarios was included. 

The **exp2flux** is available through CRAN repositories, to install it just type:
```{r, eval = FALSE}
install.packages("exp2flux")
```

```{r}
library(exp2flux)
library(sybil)
library(glpkAPI)
library(ggplot2)
```

```{r}
data("Ec_core")
Ec_core
```

```{r}
sybil::optimizeProb(object = Ec_core)
```

```{r}
data("antibioticsExpression")
names(antibioticsExpression)
```

```{r}
antibioticsExpression$`Untreated WT E coli cells at 60 min`
```

```{r}
WTmodel <- exp2flux(model = Ec_core,
                    expression = antibioticsExpression$`Untreated WT E coli cells at 60 min`,
                    outputID = "WT")
sybil::optimizeProb(WTmodel)
```

```{r}
AMPmodel <- exp2flux(model = Ec_core, 
                     expression = antibioticsExpression$`Amp treated WT E coli cells at 60 min`, 
                     outputID = "AMP")
sybil::optimizeProb(AMPmodel)
```

```{r}
head(getFluxes(model = WTmodel), 5)
head(getFluxes(model = AMPmodel), 5)
```

```{r}
head(getDifferences(model1 = WTmodel, model2 = AMPmodel), 10)
```

```{r}
plotDifferences(model1 = WTmodel, model2 = AMPmodel)
```